### PR TITLE
Kubedns logging

### LINKED ISSUE #29054: kubedns logging is missing important items
I queried a domain I KNOW has not been queried in my cluster and I don't see it in the logs.

### PR DESCRIPTION
**Pull request by bowei**
fixes 
https://github.com/kubernetes/kubernetes/issues/29053

may resolve https://github.com/kubernetes/kubernetes/issues/29054, but depends on what the specific ask is

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36013)

### CODE REVIEW

**Comment by bowei** [comment]
@thockin

**Comment by thockin** [comment]
I took a quick look.  Assigned to @MrHohn.

Can you post a snippet of logs here?

**Comment by bowei** [comment]
## v(2)

````
I1102 02:49:28.492030       1 dns.go:42] version: v1.5.0-alpha.2.320+c45624e3473292-dirty
I1102 02:49:28.492372       1 server.go:94] Using https://10.0.0.1:443 for kubernetes master, kubernetes API: <nil>
I1102 02:49:28.499400       1 server.go:100] FLAG: --alsologtostderr="false"
I1102 02:49:28.499420       1 server.go:100] FLAG: --dns-bind-address="0.0.0.0"
...
I1102 02:49:28.499498       1 server.go:100] FLAG: --vmodule=""
I1102 02:49:28.499663       1 server.go:142] Starting SkyDNS server (0.0.0.0:10053)
I1102 02:49:28.499961       1 server.go:152] Skydns metrics enabled (/metrics:10055)
I1102 02:49:28.499976       1 dns.go:144] Starting endpointsController
I1102 02:49:28.499980       1 dns.go:147] Starting serviceController
I1102 02:49:28.499982       1 dns.go:157] Waiting for Kubernetes service
I1102 02:49:28.499986       1 dns.go:163] Waiting for service: default/kubernetes
I1102 02:49:28.500056       1 logs.go:41] skydns: ready for queries on cluster.local. for tcp://0.0.0.0:10053 [rcache 0]
I1102 02:49:28.500064       1 logs.go:41] skydns: ready for queries on cluster.local. for udp://0.0.0.0:10053 [rcache 0]
I1102 02:49:28.530095       1 dns.go:246] newService default-http-backend
I1102 02:49:28.530320       1 dns.go:356] Added SRV record &{Host:default-http-backend.kube-system.svc.cluster.local. Port:80 Priority:10 Weight:10 Text: Mail:false Ttl:30 TargetStrip:0 Group: Key:}
I1102 02:49:28.530430       1 dns.go:246] newService kubernetes-dashboard
I1102 02:49:28.530517       1 dns.go:246] newService kube-dns
I1102 02:49:28.530646       1 dns.go:356] Added SRV record &{Host:kube-dns.kube-system.svc.cluster.local. Port:53 Priority:10 Weight:10 Text: Mail:false Ttl:30 TargetStrip:0 Group: Key:}
I1102 02:49:28.530732       1 dns.go:356] Added SRV record &{Host:kube-dns.kube-system.svc.cluster.local. Port:53 Priority:10 Weight:10 Text: Mail:false Ttl:30 TargetStrip:0 Group: Key:}
...
I1102 02:49:28.531267       1 dns.go:356] Added SRV record &{Host:monitoring-influxdb.kube-system.svc.cluster.local. Port:8083 Priority:10 Weight:10 Text: Mail:false Ttl:30 TargetStrip:0 Group: Key:}
I1102 02:49:28.531393       1 dns.go:356] Added SRV record &{Host:monitoring-influxdb.kube-system.svc.cluster.local. Port:8086 Priority:10 Weight:10 Text: Mail:false Ttl:30 TargetStrip:0 Group: Key:}
I1102 02:49:28.532949       1 server.go:113] Setting up Healthz Handler (/readiness)
I1102 02:49:28.533025       1 server.go:118] Setting up cache handler (/cache)
I1102 02:49:28.533061       1 server.go:107] Status HTTP port 8081
I1102 02:54:28.500594       1 dns.go:246] newService monitoring-influxdb
I1102 02:54:28.500713       1 dns.go:356] Added SRV record &{Host:monitoring-influxdb.kube-system.svc.cluster.local. Port:8083 Priority:10 Weight:10 Text: Mail:false Ttl:30 TargetStrip:0 Group: Key:}
I1102 02:54:28.500743       1 dns.go:356] Added SRV record &{Host:monitoring-influxdb.kube-system.svc.cluster.local. Port:8086 Priority:10 Weight:10 Text: Mail:false Ttl:30 TargetStrip:0 Group: Key:}
I1102 02:54:28.500770       1 dns.go:246] newService default-http-backend
I1102 02:54:28.500792       1 dns.go:356] Added SRV record &{Host:default-http-backend.kube-system.svc.cluster.local. Port:80 Priority:10 Weight:10 Text: Mail:false Ttl:30 TargetStrip:0 Group: Key:}
...
I1102 02:54:28.500884       1 dns.go:356] Added SRV record &{Host:kube-dns.kube-system.svc.cluster.local. Port:53 Priority:10 Weight:10 Text: Mail:false Ttl:30 TargetStrip:0 Group: Key:}
I1102 02:54:28.500915       1 dns.go:246] newService kubernetes
I1102 02:54:28.500935       1 dns.go:356] Added SRV record &{Host:kubernetes.default.svc.cluster.local. Port:443 Priority:10 Weight:10 Text: Mail:false Ttl:30 TargetStrip:0 Group: Key:}
````

## v(3)

````
I1102 06:10:49.747533       1 dns.go:42] version: v1.5.0-alpha.2.320+c45624e3473292-dirty
I1102 06:10:49.748365       1 server.go:94] Using https://10.0.0.1:443 for kubernetes master, kubernetes API: <nil>
I1102 06:10:49.750104       1 server.go:100] FLAG: --alsologtostderr="false"
...
I1102 06:10:49.751432       1 server.go:100] FLAG: --vmodule=""
I1102 06:10:49.751738       1 server.go:142] Starting SkyDNS server (0.0.0.0:10053)
I1102 06:10:49.753805       1 server.go:152] Skydns metrics enabled (/metrics:10055)
I1102 06:10:49.753894       1 dns.go:144] Starting endpointsController
I1102 06:10:49.753952       1 dns.go:147] Starting serviceController
I1102 06:10:49.754004       1 dns.go:157] Waiting for Kubernetes service
I1102 06:10:49.754060       1 dns.go:163] Waiting for service: default/kubernetes
I1102 06:10:49.755078       1 logs.go:41] skydns: ready for queries on cluster.local. for tcp://0.0.0.0:10053 [rcache 0]
I1102 06:10:49.757299       1 logs.go:41] skydns: ready for queries on cluster.local. for udp://0.0.0.0:10053 [rcache 0]
I1102 06:10:49.756394       1 reflector.go:202] Starting reflector *api.Endpoints (5m0s) from pkg/dns/dns.go:145
I1102 06:10:49.757653       1 reflector.go:240] Listing and watching *api.Endpoints from pkg/dns/dns.go:145
I1102 06:10:49.756580       1 reflector.go:202] Starting reflector *api.Service (5m0s) from pkg/dns/dns.go:148
I1102 06:10:49.758172       1 reflector.go:240] Listing and watching *api.Service from pkg/dns/dns.go:148
I1102 06:10:49.784963       1 dns.go:246] newService kube-dns
I1102 06:10:49.785151       1 dns.go:356] Added SRV record &{Host:kube-dns.kube-system.svc.cluster.local. Port:53 Priority:10 Weight:10 Text: Mail:false Ttl:30 TargetStrip:0 Group: Key:}
...
I1102 06:10:51.326316       1 dns.go:487] Query for "kubernetes.default.svc.cluster.local.", exact: false
I1102 06:10:51.326366       1 dns.go:603] Found 1 records for [local cluster svc default kubernetes] in the cache
I1102 06:10:51.326651       1 dns.go:487] Query for "kubernetes.default.svc.cluster.local.", exact: false
I1102 06:10:51.326667       1 dns.go:603] Found 1 records for [local cluster svc default kubernetes] in the cache
I1102 06:10:51.326976       1 dns.go:487] Query for "kubernetes.default.svc.cluster.local.", exact: false
````

## v(4)

````
I1102 06:19:26.556571       1 dns.go:487] Query for "kubernetes.default.svc.cluster.local.", exact: false
I1102 06:19:26.556582       1 dns.go:487] Query for "kubernetes.default.svc.cluster.local.", exact: false
I1102 06:19:26.556598       1 dns.go:712] not a federation query: len(["kubernetes" "default" "svc" "cluster" "local"]) != 4+len(["local" "cluster"])
I1102 06:19:26.556606       1 dns.go:712] not a federation query: len(["kubernetes" "default" "svc" "cluster" "local"]) != 4+len(["local" "cluster"])
I1102 06:19:26.556620       1 dns.go:603] Found 1 records for [local cluster svc default kubernetes] in the cache
I1102 06:19:26.556626       1 dns.go:603] Found 1 records for [local cluster svc default kubernetes] in the cache
I1102 06:19:26.556628       1 dns.go:610] getRecordsForPath retval=[{Host:10.0.0.1 Port:0 Priority:10 Weight:10 Text: Mail:false Ttl:30 TargetStrip:0 Group: Key:/skydns/local/cluster/svc/default/kubernetes/3234633364383235}], path=[local cluster svc default kubernetes]
I1102 06:19:26.556634       1 dns.go:610] getRecordsForPath retval=[{Host:10.0.0.1 Port:0 Priority:10 Weight:10 Text: Mail:false Ttl:30 TargetStrip:0 Group: Key:/skydns/local/cluster/svc/default/kubernetes/3234633364383235}], path=[local cluster svc default kubernetes]
I1102 06:19:26.556656       1 dns.go:516] Records for kubernetes.default.svc.cluster.local.: [{10.0.0.1 0 10 10  false 30 0  /skydns/local/cluster/svc/default/kubernetes/3234633364383235}]
I1102 06:19:26.556645       1 dns.go:516] Records for kubernetes.default.svc.cluster.local.: [{10.0.0.1 0 10 10  false 30 0  /skydns/local/cluster/svc/default/kubernetes/3234633364383235}]
I1102 06:19:26.557034       1 dns.go:487] Query for "kubernetes.default.svc.cluster.local.", exact: false
I1102 06:19:26.557076       1 dns.go:712] not a federation query: len(["kubernetes" "default" "svc" "cluster" "local"]) != 4+len(["local" "cluster"])
I1102 06:19:26.557090       1 dns.go:603] Found 1 records for [local cluster svc default kubernetes] in the cache
I1102 06:19:26.557096       1 dns.go:610] getRecordsForPath retval=[{Host:10.0.0.1 Port:0 Priority:10 Weight:10 Text: Mail:false Ttl:30 TargetStrip:0 Group: Key:/skydns/local/cluster/svc/default/kubernetes/3234633364383235}], path=[local cluster svc default kubernetes]
I1102 06:19:26.557107       1 dns.go:516] Records for kubernetes.default.svc.cluster.local.: [{10.0.0.1 0 10 10  false 30 0  /skydns/local/cluster/svc/default/kubernetes/3234633364383235}]
I1102 06:19:26.557704       1 dns.go:487] Query for "kubernetes.default.svc.cluster.local.", exact: false
I1102 06:19:26.557778       1 dns.go:712] not a federation query: len(["kubernetes" "default" "svc" "cluster" "local"]) != 4+len(["local" "cluster"])
I1102 06:19:26.557853       1 dns.go:603] Found 1 records for [local cluster svc default kubernetes] in the cache
I1102 06:19:26.557898       1 dns.go:610] getRecordsForPath retval=[{Host:10.0.0.1 Port:0 Priority:10 Weight:10 Text: Mail:false Ttl:30 TargetStrip:0 Group: Key:/skydns/local/cluster/svc/default/kubernetes/3234633364383235}], path=[local cluster svc default kubernetes]
I1102 06:19:26.557972       1 dns.go:516] Records for kubernetes.default.svc.cluster.local.: [{10.0.0.1 0 10 10  false 30 0  /skydns/local/cluster/svc/default/kubernetes/3234633364383235}]
````

File: pkg/dns/treecache/treecache_test.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+package treecache
+
+import (
+	"testing"
+
+	"github.com/skynetservices/skydns/msg"
+)
+
+func TestTreeCache(t *testing.T) {
```
Comment (MrHohn) [inline pkg/dns/treecache/treecache_test.go]: Should we also test SetSubCache() here?

Saw the TODO says we should combine it with SetEntry() in future. May be wait til then?

File: pkg/dns/dns.go [hunk 2]
```diff
@@ -556,25 +588,26 @@ func (kd *KubeDNS) getRecordsForPath(path []string, exact bool) ([]skymsg.Servic
...
+	records := kd.cache.GetValuesForPathWithWildcards(path...)
+	glog.V(3).Infof("Found %d records for %v in the cache", len(records), path)
 
 	retval := []skymsg.Service{}
 	for _, val := range records {
 		retval = append(retval, *val)
 	}
 
-	glog.V(2).Infof("records:%v, retval:%v, path:%v", records, retval, path)
+	glog.V(4).Infof("getRecordsForPath retval=%+v, path=%v", retval, path)
```
Comment (MrHohn) [inline pkg/dns/dns.go]: I guess it also make sense to make identical to the function name as it is V4?

File: pkg/dns/dns.go [hunk 3]
```diff
@@ -676,34 +709,39 @@ func (kd *KubeDNS) getPodIP(path []string) (string, error) {
 //   We can add support for wildcard queries later, if needed.
 func (kd *KubeDNS) isFederationQuery(path []string) bool {
 	if len(path) != 4+len(kd.domainPath) {
-		glog.V(2).Infof("not a federation query: len(%q) != 4+len(%q)", path, kd.domainPath)
+		glog.V(4).Infof("not a federation query: len(%q) != 4+len(%q)", path, kd.domainPath)
```
Comment (MrHohn) [inline pkg/dns/dns.go]: nit: first letter capital

File: pkg/dns/treecache/treecache_test.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+	tc := NewTreeCache()
+
+	{
```
Comment (MrHohn) [inline pkg/dns/treecache/treecache_test.go]: Why put this in a block?

File: pkg/dns/treecache/treecache_test.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+	// Delete some paths
+	if !tc.DeletePath("p1", "p2") {
+		t.Fatal("should delete path p2.p1.")
+	}
+	if _, ok := tc.GetEntry("key3", "p1", "p3"); !ok {
+		t.Error("should not affect p3.p1.")
+	}
+	if tc.DeletePath("p1", "p2") {
+		t.Fatal()
```
Comment (MrHohn) [inline pkg/dns/treecache/treecache_test.go]: May be add a bit error message when fail?

File: pkg/dns/treecache/treecache_test.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+	for _, testCase := range []struct {
+		k string
+		p []string
+	}{
+		{"key1", []string{"p1", "p2"}},
+		{"key2", []string{"p1", "p2"}},
+		{"key3", []string{"p1", "p3"}},
+	} {
+		if _, ok := tc.GetEntry(testCase.k, testCase.p...); ok {
+			t.Error()
```
Comment (MrHohn) [inline pkg/dns/treecache/treecache_test.go]: Should have error message.

**Review by MrHohn — COMMENTED** [review]
Looks good. Add  comments basically for nits.

File: cmd/kube-dns/app/server.go [hunk 7]
```diff
@@ -126,25 +127,30 @@ func (server *KubeDNSServer) setupHealthzHandlers() {
...
+	skydnsConfig := &server.Config{
+		Domain:  d.domain,
+		DnsAddr: fmt.Sprintf("%s:%d", d.dnsBindAddress, d.dnsPort),
+	}
 	server.SetDefaults(skydnsConfig)
 	s := server.New(d.kd, skydnsConfig)
 	if err := metrics.Metrics(); err != nil {
-		glog.Fatalf("skydns: %s", err)
+		glog.Fatalf("Skydns metrics error: %s", err)
+	} else if metrics.Port != "" {
```
Comment (bowei) [inline cmd/kube-dns/app/server.go]: ok -- added info log

File: pkg/dns/dns.go [hunk 8]
```diff
@@ -230,7 +243,9 @@ func assertIsService(obj interface{}) (*kapi.Service, bool) {
 
 func (kd *KubeDNS) newService(obj interface{}) {
 	if service, ok := assertIsService(obj); ok {
-		glog.V(4).Infof("Add/Updated for service %v", service.Name)
+		glog.V(2).Infof("newService %v", service.Name)
```
Comment (bowei) [inline pkg/dns/dns.go]: done

File: pkg/dns/dns.go (hunk 2 above)
Comment (bowei) [inline pkg/dns/dns.go]: it is the same?

File: pkg/dns/dns.go (hunk 3 above)
Comment (bowei) [inline pkg/dns/dns.go]: done

File: pkg/dns/util/util.go [hunk 9]
```diff
@@ -63,7 +63,8 @@ func ReverseArray(arr []string) []string {
 func GetSkyMsg(ip string, port int) (*msg.Service, string) {
 	msg := NewServiceRecord(ip, port)
 	hash := HashServiceRecord(msg)
-	glog.V(2).Infof("DNS Record:%s, hash:%s", fmt.Sprintf("%v", msg), hash)
+	glog.V(5).Infof("Created new DNS record: %s, hash:%s",
```
Comment (bowei) [inline pkg/dns/util/util.go]: changed to "constructed"

File: pkg/dns/treecache/treecache_test.go (hunk 4 above)
Comment (bowei) [inline pkg/dns/treecache/treecache_test.go]: limit variable scoping

File: pkg/dns/treecache/treecache_test.go (hunk 1 above)
Comment (bowei) [inline pkg/dns/treecache/treecache_test.go]: add a test

File: pkg/dns/treecache/treecache_test.go (hunk 5 above)
Comment (bowei) [inline pkg/dns/treecache/treecache_test.go]: ok

File: pkg/dns/treecache/treecache_test.go (hunk 6 above)
Comment (bowei) [inline pkg/dns/treecache/treecache_test.go]: ok

**Comment by MrHohn** [comment]
@k8s-bot gci gce e2e test this, issue: #34665

**Comment by bowei** [comment]
gci gce e2e test this

**Comment by bowei** [comment]
@k8s-bot gci gce e2e test this

**Pull request merged**